src/hotspot/share/nmt/virtualMemoryTracker.cpp

```diff
@@ -47,545 +46,213 @@ void VirtualMemory::update_peak(size_t size) {
-    if (reserved_rgn->mem_tag() != mem_tag) {
-      assert(reserved_rgn->mem_tag() == mtNone, "Overwrite memory tag (should be mtNone, is: \"%s\")",
-             NMTUtil::tag_to_name(reserved_rgn->mem_tag()));
-      reserved_rgn->set_mem_tag(mem_tag);
-    }
-  }
+bool VirtualMemoryTracker::walk_virtual_memory(VirtualMemoryWalker* walker) {
+  MemTracker::NmtVirtualMemoryLocker nvml;
+  tree()->visit_reserved_regions([&](ReservedMemoryRegion& rgn) {
+    log_info(nmt)("region in walker vmem, base: " INTPTR_FORMAT " size: %zu , %s, committed: %zu",
+     p2i(rgn.base()), rgn.size(), rgn.tag_name(), rgn.committed_size());
+    if (!walker->do_allocation_site(&rgn))
+      return false;
+    return true;
+  });
+  return true;
 }
 
-bool VirtualMemoryTracker::add_committed_region(address addr, size_t size,
-  const NativeCallStack& stack) {
-  assert(addr != nullptr, "Invalid address");
-  assert(size > 0, "Invalid size");
-  assert(_reserved_regions != nullptr, "Sanity check");
-  MemTracker::assert_locked();
-
-  ReservedMemoryRegion  rgn(addr, size);
-  ReservedMemoryRegion* reserved_rgn = _reserved_regions->find(rgn);
-
-  if (reserved_rgn == nullptr) {
-    log_debug(nmt)("Add committed region \'%s\', No reserved region found for  (" INTPTR_FORMAT ", %zu)",
-                  rgn.mem_tag_name(),  p2i(rgn.base()), rgn.size());
-  }
-  assert(reserved_rgn != nullptr, "Add committed region, No reserved region found");
-  assert(reserved_rgn->contain_region(addr, size), "Not completely contained");
-  bool result = reserved_rgn->add_committed_region(addr, size, stack);
-  log_debug(nmt)("Add committed region \'%s\'(" INTPTR_FORMAT ", %zu) %s",
-                reserved_rgn->mem_tag_name(),  p2i(rgn.base()), rgn.size(), (result ? "Succeeded" : "Failed"));
-  return result;
+int compare_committed_region(const CommittedMemoryRegion& r1, const CommittedMemoryRegion& r2) {
+  return r1.compare(r2);
 }
 
-bool VirtualMemoryTracker::remove_uncommitted_region(address addr, size_t size) {
-  assert(addr != nullptr, "Invalid address");
-  assert(size > 0, "Invalid size");
-  assert(_reserved_regions != nullptr, "Sanity check");
-  MemTracker::assert_locked();
-
-  ReservedMemoryRegion  rgn(addr, size);
-  ReservedMemoryRegion* reserved_rgn = _reserved_regions->find(rgn);
-  assert(reserved_rgn != nullptr, "No reserved region (" INTPTR_FORMAT ", %zu)", p2i(addr), size);
-  assert(reserved_rgn->contain_region(addr, size), "Not completely contained");
-  const char* type_name = reserved_rgn->mem_tag_name();  // after remove, info is not complete
-  bool result = reserved_rgn->remove_uncommitted_region(addr, size);
-  log_debug(nmt)("Removed uncommitted region \'%s\' (" INTPTR_FORMAT ", %zu) %s",
-                 type_name,  p2i(addr), size, (result ? " Succeeded" : "Failed"));
-  return result;
+int compare_reserved_region_base(const ReservedMemoryRegion& r1, const ReservedMemoryRegion& r2) {
+  return r1.compare(r2);
```

Comment:
Why did we name it `compare_reserved_region_base`, not simply `compare_reserved_region`